x-pack/legacy/plugins/fleet/server/libs/agent.ts

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import moment from 'moment';
+import {
+  AgentAdapter,
+  Agent,
+  SortOptions,
+  NewAgent,
+  AgentType,
+} from './adapters/agent/adapter_type';
+import { TokenLib } from './token';
+
+export class AgentLib {
+  constructor(private readonly agentAdater: AgentAdapter, private readonly tokens: TokenLib) {}
+
+  /**
+   * Enroll a new token into elastic fleet
+   */
+  public async enroll(
+    token: any,
+    type: AgentType,
+    config: { id: string; sharedId: string },
+    metadata?: { local: any; userProvided: any },
+    sharedId?: string
+  ): Promise<Agent> {
+    const verifyResponse = await this.tokens.verify(token);
+    if (!verifyResponse.valid) {
+      throw new Error(`Enrollment token is not valid: ${verifyResponse.reason}`);
+    }
+
+    const existingAgent = sharedId ? await this.agentAdater.getBySharedId(sharedId) : null;
+
+    if (existingAgent && existingAgent.active === true) {
+      throw new Error('Impossible to enroll an already active agent');
+    }
+
+    const accessToken = await this.tokens.generateAccessToken(token);
+    const enrolledAt = moment().toISOString();
+
+    const parentId =
+      type === 'EPHEMERAL_INSTANCE'
+        ? (await this._findOrCreateParentForEphemeral(config.id, config.sharedId)).id
```

Comment:
I saw a potential race condition here, what mechanism can we use here to ensure we create only one parent for ephemeral?

Reply:
Sounds a good solution, going to look a little to SO to see what is possible to do here

Reply:
Looks like we can do an overwrite and specify the id in the create option, I should be able to use that